## Cleaning Process Lesson
### Gather

In [1]:
import pandas as pd
import numpy as np

In [2]:
patients = pd.read_csv('patients.csv')
treatments = pd.read_csv('treatments.csv')
adverse_reactions = pd.read_csv('adverse_reactions.csv')

### Assess
These are the programmatic assessment methods in pandas that you will probably use most often:

* .head (DataFrame and Series)
* .tail (DataFrame and Series)
* .sample (DataFrame and Series)
* .info (DataFrame only)
* .describe (DataFrame and Series)
* .value_counts (Series only)
* Various methods of indexing and selecting data (.loc and bracket notation with/without boolean indexing, also .iloc)

Try `.head` and `.tail` on the `patients` table.

##### patients table

In [3]:
patients.head(3)

patient_id assigned_sex given_name  surname                    address  \
0           1       female        Zoe  Wellish       576 Brown Bear Drive   
1           2       female     Pamela     Hill  2370 University Hill Road   
2           3         male        Jae   Debord      1493 Poling Farm Road   

                city       state  zip_code        country  \
0  Rancho California  California   92390.0  United States   
1          Armstrong    Illinois   61812.0  United States   
2               York    Nebraska   68467.0  United States   

                                 contact  birthdate  weight  height   bmi  
0   951-719-9170ZoeWellish@superrito.com  7/10/1976   121.7      66  19.6  
1  PamelaSHill@cuvox.de+1 (217) 569-3204   4/3/1967   118.8      66  19.2  
2       402-363-6804JaeMDebord@gustr.com  2/19/1980   177.8      71  24.8

In [4]:
patients.tail(3)

patient_id assigned_sex given_name      surname  \
500         501       female      Jinke    de Keizer   
501         502       female    Chidalu  Onyekaozulu   
502         503         male        Pat      Gersten   

                      address           city     state  zip_code  \
500         649 Nutter Street  Overland Park        MO   64110.0   
501  3652 Boone Crockett Lane        Seattle        WA   98109.0   
502         2778 North Avenue           Burr  Nebraska   68324.0   

           country                                          contact  \
500  United States            816-223-6007JinkedeKeizer@teleworm.us   
501  United States  ChidaluOnyekaozulu@jourrapide.com1 360 443 2060   
502  United States             PatrickGersten@rhyta.com402-848-4923   

     birthdate  weight  height   bmi  
500  1/13/1971   171.2      67  26.8  
501  2/13/1952   176.9      67  27.7  
502   5/3/1954   138.2      71  19.3

In [5]:
patients.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 503 entries, 0 to 502
Data columns (total 14 columns):
patient_id      503 non-null int64
assigned_sex    503 non-null object
given_name      503 non-null object
surname         503 non-null object
address         491 non-null object
city            491 non-null object
state           491 non-null object
zip_code        491 non-null float64
country         491 non-null object
contact         491 non-null object
birthdate       503 non-null object
weight          503 non-null float64
height          503 non-null int64
bmi             503 non-null float64
dtypes: float64(3), int64(2), object(9)
memory usage: 55.1+ KB


In [6]:
patients.describe()

patient_id      zip_code      weight      height         bmi
count  503.000000    491.000000  503.000000  503.000000  503.000000
mean   252.000000  49084.118126  173.434990   66.634195   27.483897
std    145.347859  30265.807442   33.916741    4.411297    5.276438
min      1.000000   1002.000000   48.800000   27.000000   17.100000
25%    126.500000  21920.500000  149.300000   63.000000   23.300000
50%    252.000000  48057.000000  175.300000   67.000000   27.200000
75%    377.500000  75679.000000  199.500000   70.000000   31.750000
max    503.000000  99701.000000  255.900000   79.000000   37.700000

In [7]:
patients.sample(5)

patient_id assigned_sex given_name     surname               address  \
191         192       female    Fanette    Fournier  602 Tator Patch Road   
342         343       female   Mathilde    Nørgaard   846 Copperhead Road   
105         106         male       Nate      Bowden   4310 Johnson Street   
288         289         male     Fareeq        Kalb    1043 Ventura Drive   
8             9         male      Dsvid  Gustafsson    1790 Nutter Street   

            city       state  zip_code        country  \
191      Chicago    Illinois   60607.0  United States   
342     Hartford          CT    6182.0  United States   
105      Clinton          NC   28328.0  United States   
288   Santa Cruz  California   95060.0  United States   
8    Kansas City          MO   64105.0  United States   

                                          contact   birthdate  weight  height  \
191   FanetteFournier@einrot.com+1 (312) 864-5737  10/24/1938   112.6      60   
342  MathildeDNrgaard@superrito.com1 860 618 4560   1/21/1985   127.2      61   
105             NateBowden@einrot.com919-915-8670   3/26/1954   222.0      71   
288      831-427-4114FareeqJawharKalb@armyspy.com    5/1/1985   194.5      70   
8         816-265-9578DavidGustafsson@armyspy.com    3/6/1937   163.9      66   

      bmi  
191  22.0  
342  24.0  
105  31.0  
288  27.9  
8    26.5

In [8]:
patients.surname.value_counts()

Doe            6
Taylor         3
Jakobsen       3
Lâm            2
Schiavone      2
Ogochukwu      2
Kowalczyk      2
Parker         2
Johnson        2
Liễu           2
Correia        2
Woźniak        2
Tucker         2
Kadyrov        2
Collins        2
Cabrera        2
Berg           2
Lương          2
Aranda         2
Lund           2
Dratchev       2
Souza          2
Silva          2
Bùi            2
Hueber         2
Grímsdóttir    2
Nilsen         2
Batukayev      2
Cindrić        2
Tạ             2
              ..
Cumpston       1
Nakahara       1
McIntosh       1
Balagić        1
Addison        1
Day            1
Scheltens      1
Vestergaard    1
Beauvais       1
Allen          1
Cotton         1
Schneider      1
McGregor       1
Šlezinger      1
Azevedo        1
Lê             1
Csonka         1
Galić          1
Yao            1
Osman          1
Khoury         1
Enoksen        1
Achebe         1
Györfy         1
Prince         1
Pinto          1
Nyborg         1
Shishani      

In [9]:
patients.query('given_name == "Jake" or surname == "Jakobsen"')

patient_id assigned_sex given_name   surname             address  \
24           25         male      Jakob  Jakobsen   648 Old Dear Lane   
29           30         male       Jake  Jakobsen   648 Old Dear Lane   
432         433       female      Karen  Jakobsen  1690 Fannie Street   

            city     state  zip_code        country  \
24   Port Jervis  New York   12771.0  United States   
29   Port Jervis  New York   12771.0  United States   
432      Houston        TX   77020.0  United States   

                                        contact   birthdate  weight  height  \
24   JakobCJakobsen@einrot.com+1 (845) 858-7707    8/1/1985   155.8      67   
29   JakobCJakobsen@einrot.com+1 (845) 858-7707    8/1/1985   155.8      67   
432  KarenJakobsen@jourrapide.com1 979 203 0438  11/25/1962   185.2      67   

      bmi  
24   24.4  
29   24.4  
432  29.0

In [10]:
patients.query('given_name == "Elizabeth" or surname == "Knudsen"')

patient_id assigned_sex given_name  surname address city state  zip_code  \
230         231       female  Elisabeth  Knudsen     NaN  NaN   NaN       NaN   

    country contact  birthdate  weight  height   bmi  
230     NaN     NaN  9/23/1976   165.9      63  29.4

In [11]:
patients.address.duplicated().sum()

19

In [12]:
patients.address.value_counts()

123 Main Street               6
648 Old Dear Lane             2
2476 Fulton Street            2
2778 North Avenue             2
4605 Hall Street              1
1386 Farland Street           1
1233 Liberty Avenue           1
1009 Felosa Drive             1
2607 Water Street             1
3730 Seth Street              1
2310 Overlook Drive           1
2324 Benson Street            1
3743 Wescam Court             1
4852 Rose Avenue              1
846 Copperhead Road           1
4111 Thunder Road             1
1763 Williams Mine Road       1
2103 Edington Drive           1
1934 August Lane              1
4458 Stark Hollow Road        1
707 Gateway Avenue            1
2884 Elsie Drive              1
353 Whaley Lane               1
1346 Nicholas Street          1
1341 Chatham Way              1
2785 Norma Lane               1
1865 Honeysuckle Lane         1
2690 Pin Oak Drive            1
1525 Crestview Terrace        1
4438 Reeves Street            1
                             ..
208 Birc

In [13]:
patients.weight.sort_values()

210     48.8
459    102.1
335    102.7
74     103.2
317    106.0
171    106.5
51     107.1
270    108.1
198    108.5
48     109.1
478    109.6
141    110.2
38     111.8
438    112.0
14     112.0
235    112.2
307    112.4
191    112.6
408    113.1
49     113.3
326    114.0
338    114.1
253    117.0
321    118.4
168    118.8
1      118.8
350    119.0
207    119.2
265    120.0
341    120.3
       ...  
332    224.0
252    224.2
12     224.2
222    224.8
166    225.3
111    225.9
101    226.2
150    226.6
352    227.7
428    227.7
88     227.7
13     228.4
339    229.0
182    230.3
121    230.8
257    231.7
395    231.9
246    232.1
219    237.8
11     238.7
50     238.9
441    239.1
499    239.6
439    242.0
487    242.4
144    244.9
61     244.9
283    245.5
118    254.5
485    255.9
Name: weight, Length: 503, dtype: float64

In [14]:
weight_lbs = patients[patients.surname == 'Zaitseva'].weight * 2.20462
height_in = patients[patients.surname == 'Zaitseva'].height
bmi_check = 703 * weight_lbs / (height_in * height_in)
bmi_check

210    19.055827
dtype: float64

In [15]:
patients[patients.surname == 'Zaitseva'].bmi

210    19.1
Name: bmi, dtype: float64

In [16]:
patients[patients['city'] == 'New York'].shape

(18, 14)

In [17]:
patients[patients['address'].isnull()]

patient_id assigned_sex given_name      surname address city state  \
209         210       female     Lalita  Eldarkhanov     NaN  NaN   NaN   
219         220         male         Mỹ        Quynh     NaN  NaN   NaN   
230         231       female  Elisabeth      Knudsen     NaN  NaN   NaN   
234         235       female    Martina    Tománková     NaN  NaN   NaN   
242         243         male       John      O'Brian     NaN  NaN   NaN   
249         250         male   Benjamin       Mehler     NaN  NaN   NaN   
257         258         male        Jin         Kung     NaN  NaN   NaN   
264         265       female   Wafiyyah       Asfour     NaN  NaN   NaN   
269         270       female     Flavia   Fiorentino     NaN  NaN   NaN   
278         279       female   Generosa        Cabán     NaN  NaN   NaN   
286         287         male      Lewis         Webb     NaN  NaN   NaN   
296         297       female        Chỉ          Lâm     NaN  NaN   NaN   

     zip_code country contact   birthdate  weight  height   bmi  
209       NaN     NaN     NaN   8/14/1950   143.4      62  26.2  
219       NaN     NaN     NaN    4/9/1978   237.8      69  35.1  
230       NaN     NaN     NaN   9/23/1976   165.9      63  29.4  
234       NaN     NaN     NaN    4/7/1936   199.5      65  33.2  
242       NaN     NaN     NaN   2/25/1957   205.3      74  26.4  
249       NaN     NaN     NaN  10/30/1951   146.5      69  21.6  
257       NaN     NaN     NaN   5/17/1995   231.7      69  34.2  
264       NaN     NaN     NaN   11/3/1989   158.6      63  28.1  
269       NaN     NaN     NaN   10/9/1937   175.2      61  33.1  
278       NaN     NaN     NaN  12/16/1962   124.3      69  18.4  
286       NaN     NaN     NaN    4/1/1979   155.3      68  23.6  
296       NaN     NaN     NaN   5/14/1990   181.1      63  32.1

Try `.sample` on the `treatments` table.

##### treatments

In [18]:
treatments.sample(3)

given_name      surname    auralin    novodra  hba1c_start  hba1c_end  \
40        ásta  grímsdóttir          -  29u - 30u         7.62       7.16   
190     martha   afanasyeva          -  34u - 41u         7.81       7.39   
196      peter     pospíšil  51u - 60u          -         7.76       7.34   

     hba1c_change  
40           0.96  
190           NaN  
196           NaN

Try `.info` on the `treatments` table.

In [19]:
treatments.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 280 entries, 0 to 279
Data columns (total 7 columns):
given_name      280 non-null object
surname         280 non-null object
auralin         280 non-null object
novodra         280 non-null object
hba1c_start     280 non-null float64
hba1c_end       280 non-null float64
hba1c_change    171 non-null float64
dtypes: float64(3), object(4)
memory usage: 15.4+ KB


Try `.describe` on the `patients` table.

In [20]:
treatments.describe()

hba1c_start   hba1c_end  hba1c_change
count   280.000000  280.000000    171.000000
mean      7.985929    7.589286      0.546023
std       0.568638    0.569672      0.279555
min       7.500000    7.010000      0.200000
25%       7.660000    7.270000      0.340000
50%       7.800000    7.420000      0.380000
75%       7.970000    7.570000      0.920000
max       9.950000    9.580000      0.990000

In [21]:
sum(treatments.auralin.isnull())

0

In [22]:
sum(treatments.novodra.isnull())

0

Try `.value_counts` on the *adverse_reaction* column of the `adverse_reactions` table.

In [23]:
adverse_reactions.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 34 entries, 0 to 33
Data columns (total 3 columns):
given_name          34 non-null object
surname             34 non-null object
adverse_reaction    34 non-null object
dtypes: object(3)
memory usage: 896.0+ bytes


In [24]:
adverse_reactions.describe()

given_name  surname adverse_reaction
count          34       34               34
unique         34       33                6
top         berta  johnson     hypoglycemia
freq            1        2               19

In [25]:
adverse_reactions.adverse_reaction.value_counts()

hypoglycemia                 19
injection site discomfort     6
headache                      3
nausea                        2
cough                         2
throat irritation             2
Name: adverse_reaction, dtype: int64

Try selecting the records in the `patients` table for patients that are from the *city* New York.

#### Quality

##### patients table

* zip code is a float not a strig
* for some registries zip code has four digits
* Tim Neudorf height is 27 instead of 71 inches
* Sometimes there's full state name and other times just the abbreviations
* patient_id 9 has name with typo: Dsvid Gustafson
* missing demographic information (address - contact columns)
* there are 4 columns with wrong data types: assigned_sex, state, zip_code and birthdate)
* different phone number formats
* John Doe is duplicated
* multiple records for Jakobsen, Gersten, Taylor
* kgs instead of lbs for Zaitseva weight

##### treatements table

* missing HbA1c changes
* the letter u in starting and ending doses for Auralin and Novodra
* lowercase given names and surnames
* missing records (280 instead of 350)
* wrong data types for auralin and novodra columns
* inaccurate value for HbA1c changes (4s mistaken by 9s)
* nulls represented as dashes `-` in auralin and novodra columns

##### adverse_reactions table

* lowercase given names and surnames

#### Tidiness

##### patients table

* `contact` column should be split into phone number and email

##### treatements table

* `auralin` and `novodra` should have two columns, one for minimum and other for maximum dose
* `auralin` and `novodra` are different types of treatment
* given name and surname columns are duplicated in treatments

##### adverse_reactions table

* this table is part of the treatment


### Clean


In [26]:
patients_clean = patients.copy()
treatments_clean = treatments.copy()
adverse_reactions_clean = adverse_reactions.copy()

### Missing Data
#### `treatments`: Missing records (280 instead of 350)
##### Define
Add missing `treatments` records from file `treatments_cut.csv` to `treatments_clean` DataFrame.

##### Code

In [27]:
treatments_cut = pd.read_csv('treatments_cut.csv')
treatments_cut.head(2)

given_name    surname    auralin novodra  hba1c_start  hba1c_end  \
0      jožka  resanovič  22u - 30u       -         7.56       7.22   
1  inunnguaq   heilmann  57u - 67u       -         7.85       7.45   

   hba1c_change  
0          0.34  
1           NaN

In [28]:
treatments_clean.head(2)

given_name     surname    auralin    novodra  hba1c_start  hba1c_end  \
0   veronika    jindrová  41u - 48u          -         7.63       7.20   
1     elliot  richardson          -  40u - 45u         7.56       7.09   

   hba1c_change  
0           NaN  
1          0.97

In [29]:
treatments_clean = pd.concat([treatments_clean, treatments_cut], ignore_index=True)

##### Test

In [30]:
assert treatments_clean.shape[0] == 350

#### `treatments`: Missing HbA1c changes and Inaccurate HbA1c changes (leading 4s mistaken as 9s)

##### Define
Calculate `HbA1c changes` by subtracting the columns `hba1c_start` and `hba1c_end`, and storing the new value in column `hba1c_change`.

##### Code

In [31]:
treatments_clean['hba1c_change'] = treatments_clean['hba1c_start'] - treatments_clean['hba1c_end']

##### Test

In [32]:
assert treatments_clean.hba1c_change.isna().sum() == 0

### Tidiness

#### Contact column in `patients` table contains two variables: phone number and email
##### Define
Split column `contact` into two columns, one for phone number and other for email.

##### Code

In [33]:
patients_clean['phone_number'] = (patients_clean.contact
                                                .str
                                                .extract(r'((?:\+\d{1,2}\s)?\(?\d{3}\)?[\s.-]?\d{3}[\s.-]?\d{4})'))

In [34]:
patients_clean['email'] = (patients_clean.contact
                                         .str
                                         .extract(r'([a-zA-Z][a-zA-Z0-9_.+-]+@[a-zA-Z0-9-]+\.[a-zA-Z0-9-.]+[a-zA-Z])'))



In [35]:
patients_clean.drop(columns=['contact'], inplace=True)

##### Test

In [36]:
patients_clean.head()

patient_id assigned_sex given_name  surname                    address  \
0           1       female        Zoe  Wellish       576 Brown Bear Drive   
1           2       female     Pamela     Hill  2370 University Hill Road   
2           3         male        Jae   Debord      1493 Poling Farm Road   
3           4         male       Liêm     Phan        2335 Webster Street   
4           5         male        Tim  Neudorf       1428 Turkey Pen Lane   

                city       state  zip_code        country  birthdate  weight  \
0  Rancho California  California   92390.0  United States  7/10/1976   121.7   
1          Armstrong    Illinois   61812.0  United States   4/3/1967   118.8   
2               York    Nebraska   68467.0  United States  2/19/1980   177.8   
3         Woodbridge          NJ    7095.0  United States  7/26/1951   220.9   
4             Dothan          AL   36303.0  United States  2/18/1928   192.3   

   height   bmi       phone_number                      email  
0      66  19.6       951-719-9170   ZoeWellish@superrito.com  
1      66  19.2  +1 (217) 569-3204       PamelaSHill@cuvox.de  
2      71  24.8       402-363-6804       JaeMDebord@gustr.com  
3      70  31.7  +1 (732) 636-8246  PhanBaLiem@jourrapide.com  
4      27  26.1       334-515-7487        TimNeudorf@cuvox.de

#### Three variables in two columns in `treatments` table (treatment, start dose and end dose)
##### Define
Create a new column `treatment` to the categories of `auralin` and `novodra`. Split the dose into `start_dose` and `end_dose` columns.

*Your definition here. Hint: use pandas' [melt function](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.melt.html) and [`str.split()` method](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.Series.str.split.html). Here is an excellent [`melt` tutorial](https://deparkes.co.uk/2016/10/28/reshape-pandas-data-with-melt/).*

In [37]:
treatments_clean.head(10)

given_name     surname    auralin    novodra  hba1c_start  hba1c_end  \
0   veronika    jindrová  41u - 48u          -         7.63       7.20   
1     elliot  richardson          -  40u - 45u         7.56       7.09   
2   yukitaka    takenaka          -  39u - 36u         7.68       7.25   
3       skye  gormanston  33u - 36u          -         7.97       7.62   
4     alissa      montez          -  33u - 29u         7.78       7.46   
5    jasmine       sykes          -  42u - 44u         7.56       7.18   
6     sophia      haugen  37u - 42u          -         7.65       7.27   
7      eddie      archer  31u - 38u          -         7.89       7.55   
8      saber      ménard          -  54u - 54u         8.08       7.70   
9       asia     woźniak  30u - 36u          -         7.76       7.37   

   hba1c_change  
0          0.43  
1          0.47  
2          0.43  
3          0.35  
4          0.32  
5          0.38  
6          0.38  
7          0.34  
8          0.38  
9          0.39

In [38]:
treatments_clean = pd.melt(treatments_clean,
                           id_vars=['given_name', 'surname', 'hba1c_start', 'hba1c_end', 'hba1c_change'],
                           var_name='treatment', value_name='dose')

In [40]:
treatments_clean = treatments_clean[treatments_clean['dose'] != '-']

In [44]:
treatments_clean.head()

given_name     surname  hba1c_start  hba1c_end  hba1c_change treatment  \
0   veronika    jindrová         7.63       7.20          0.43   auralin   
3       skye  gormanston         7.97       7.62          0.35   auralin   
6     sophia      haugen         7.65       7.27          0.38   auralin   
7      eddie      archer         7.89       7.55          0.34   auralin   
9       asia     woźniak         7.76       7.37          0.39   auralin   

        dose  
0  41u - 48u  
3  33u - 36u  
6  37u - 42u  
7  31u - 38u  
9  30u - 36u

In [58]:
treatments_clean['start_dose'], treatments_clean['end_dose'] = treatments_clean['dose'].str.split('-', 0).str
treatments_clean.drop(columns=['dose'], inplace=True)

In [59]:
treatments_clean.head()

given_name     surname  hba1c_start  hba1c_end  hba1c_change treatment  \
0   veronika    jindrová         7.63       7.20          0.43   auralin   
3       skye  gormanston         7.97       7.62          0.35   auralin   
6     sophia      haugen         7.65       7.27          0.38   auralin   
7      eddie      archer         7.89       7.55          0.34   auralin   
9       asia     woźniak         7.76       7.37          0.39   auralin   

  start_dose end_dose  
0       41u       48u  
3       33u       36u  
6       37u       42u  
7       31u       38u  
9       30u       36u

##### Test

In [60]:
treatments_clean.sample(5)

given_name   surname  hba1c_start  hba1c_end  hba1c_change treatment  \
426   noriyuki     sakai         7.58       7.16          0.42   novodra   
446  abdul-nur       isa         7.98       7.53          0.45   novodra   
547   angelico  marchesi         7.62       7.18          0.44   novodra   
143       nora    nyborg         7.83       7.48          0.35   auralin   
52          mỹ     quynh         7.61       7.16          0.45   auralin   

    start_dose end_dose  
426       32u       31u  
446       54u       50u  
547       58u       58u  
143       55u       59u  
52        57u       64u

#### Adverse reaction should be part of the `treatments` table
##### Define
Merge adverse reaction info from `adverse_reaction` table with treatments table.
##### Code

In [64]:
treatments_clean = treatments_clean.merge(adverse_reactions, on=['given_name', 'surname'], how='left')

##### Test

In [68]:
treatments_clean

given_name        surname  hba1c_start  hba1c_end  hba1c_change  \
0        veronika       jindrová         7.63       7.20          0.43   
1            skye     gormanston         7.97       7.62          0.35   
2          sophia         haugen         7.65       7.27          0.38   
3           eddie         archer         7.89       7.55          0.34   
4            asia        woźniak         7.76       7.37          0.39   
5          joseph            day         7.70       7.19          0.51   
6         roxanne      andreyeva         9.54       9.14          0.40   
7          simone   baumgaertner         7.74       7.30          0.44   
8            enco         žibrik         7.78       7.34          0.44   
9         camilla       zaitseva         7.53       7.13          0.40   
10          tekla        walczak         7.61       7.29          0.32   
11    brancaleone          russo         8.61       8.18          0.43   
12           isac           berg         9.68       9.29          0.39   
13        clinton         miller         7.79       7.40          0.39   
14         eugene        mironov         7.81       7.48          0.33   
15    szilveszter          totth         7.70       7.38          0.32   
16      alexander      mathiesen         7.96       7.55          0.41   
17            chỉ            lâm         7.68       7.24          0.44   
18      władysław      wieczorek         7.92       7.47          0.45   
19       kristján        ingason         7.92       7.57          0.35   
20         marija       grubišić         7.53       7.15          0.38   
21          sauli     koivuniemi         7.67       7.37          0.30   
22        mariana          souza         7.86       7.51          0.35   
23     kristoffer      martinsen         9.18       8.64          0.54   
24             mỹ          quynh         7.61       7.16          0.45   
25           oles        zhdanov         7.52       7.11          0.41   
26        triana.       terrazas         7.71       7.34          0.37   
27         gabryŝ    tomaszewski         7.87       7.47          0.40   
28      leixandre         alanis         7.74       7.32          0.42   
29   onyekachukwu         obinna         7.58       7.12          0.46   
..            ...            ...          ...        ...           ...   
320          jane        citizen         7.98       7.60          0.38   
321        angela     lavrentyev         7.61       7.14          0.47   
322        edelma    villalpando         7.99       7.56          0.43   
323        annika          vaara         7.73       7.34          0.39   
324         chiho           higa         7.71       7.30          0.41   
325     beatrycze        woźniak         7.54       7.17          0.37   
326      miłosław     wiśniewski         7.51       7.08          0.43   
327       firenze          fodor         7.89       7.55          0.34   
328           zoe        wellish         7.71       7.30          0.41   
329           una  traustadóttir         8.00       7.50          0.50   
330         luboš          pecha         7.79       7.45          0.34   
331         meaza         brhane         7.70       7.36          0.34   
332         adlan       shishani         7.84       7.37          0.47   
333         sofia      hermansen         8.90       8.57          0.33   
334         guðni      heimisson         7.64       7.24          0.40   
335       eufemio        rosario         7.54       7.26          0.28   
336      dalmacia         madrid         7.67       7.21          0.46   
337         daimy          tromp         9.41       8.94          0.47   
338        jeremy        montagu         7.68       7.36          0.32   
339       nebechi      ekechukwu         7.78       7.39          0.39   
340       satsita      batukayev         7.63       7.25          0.38   
341       timothy         cotton         7.92       7.52          0.40   
342       b

#### Given name and surname columns in `patients` table duplicated in `treatments` and `adverse_reactions` tables  and Lowercase given names and surnames
##### Define
Change `treatments` and `adverse_reactions` to use the patient_id from table `patients`.

*Your definition here. Hint: [tutorial](https://chrisalbon.com/python/pandas_join_merge_dataframe.html) for one function used in the solution and [tutorial](http://erikrood.com/Python_References/dropping_rows_cols_pandas.html) for another function used in the solution.*
##### Code

In [69]:
patients_clean.head()

patient_id assigned_sex given_name  surname                    address  \
0           1       female        Zoe  Wellish       576 Brown Bear Drive   
1           2       female     Pamela     Hill  2370 University Hill Road   
2           3         male        Jae   Debord      1493 Poling Farm Road   
3           4         male       Liêm     Phan        2335 Webster Street   
4           5         male        Tim  Neudorf       1428 Turkey Pen Lane   

                city       state  zip_code        country  birthdate  weight  \
0  Rancho California  California   92390.0  United States  7/10/1976   121.7   
1          Armstrong    Illinois   61812.0  United States   4/3/1967   118.8   
2               York    Nebraska   68467.0  United States  2/19/1980   177.8   
3         Woodbridge          NJ    7095.0  United States  7/26/1951   220.9   
4             Dothan          AL   36303.0  United States  2/18/1928   192.3   

   height   bmi       phone_number                      email  
0      66  19.6       951-719-9170   ZoeWellish@superrito.com  
1      66  19.2  +1 (217) 569-3204       PamelaSHill@cuvox.de  
2      71  24.8       402-363-6804       JaeMDebord@gustr.com  
3      70  31.7  +1 (732) 636-8246  PhanBaLiem@jourrapide.com  
4      27  26.1       334-515-7487        TimNeudorf@cuvox.de

In [70]:
patients.shape

(503, 14)

In [76]:
ids_and_names = patients_clean[['patient_id', 'given_name', 'surname']]
ids_and_names.head()

patient_id given_name  surname
0           1        Zoe  Wellish
1           2     Pamela     Hill
2           3        Jae   Debord
3           4       Liêm     Phan
4           5        Tim  Neudorf

In [86]:
ids_and_names.given_name = ids_and_names.given_name.str.lower()
ids_and_names.surname = ids_and_names.surname.str.lower()
ids_and_names.head()

/usr/local/lib/python3.7/site-packages/pandas/core/generic.py:5096: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  self[name] = value


patient_id given_name  surname
0         1.0        zoe  wellish
1         2.0     pamela     hill
2         3.0        jae   debord
3         4.0       liêm     phan
4         5.0        tim  neudorf

In [88]:
treatments_clean = treatments_clean.merge(ids_and_names, on=['given_name', 'surname'])
treatments_clean.drop(columns=['given_name', 'surname'], inplace=True)

##### Test

In [89]:
treatments_clean.head()

hba1c_start  hba1c_end  hba1c_change treatment start_dose end_dose  \
0         7.63       7.20          0.43   auralin       41u       48u   
1         7.97       7.62          0.35   auralin       33u       36u   
2         7.65       7.27          0.38   auralin       37u       42u   
3         7.89       7.55          0.34   auralin       31u       38u   
4         7.76       7.37          0.39   auralin       30u       36u   

  adverse_reaction  patient_id  
0              NaN       225.0  
1              NaN       242.0  
2              NaN       345.0  
3              NaN       276.0  
4              NaN        15.0

In [97]:
# Checking if patient_id is the only duplicated column
all_columns = pd.Series(list(patients_clean) + list(treatments_clean))
assert all_columns[all_columns.duplicated()].values == 'patient_id'